README.md
# HashiCorp Vault Service Broker

This repository provides an implementation of the [open service broker
API][open-broker-api] for [HashiCorp's Vault][vault]. The service broker
connects to an existing Vault cluster and can be used by multiple tenants within
Cloud Foundry.

[open-broker-api]: https://openservicebrokerapi.org/

## Getting Started

The HashiCorp Vault Service Broker does not run a Vault server for you. There is
an assumption that the Vault cluster is already setup and configured. This Vault
server does not need to be running under Cloud Foundry, OpenShift, Kubernetes,
etc, but it must be accessible from within those environments or wherever the
broker is deployed.

These getting started instructions assume the Vault server's address and token
are specified as the following environment variables:

```sh
$ export VAULT_ADDR="https://vault.company.internal/"
$ export VAULT_TOKEN="abcdef-134255-..."
```

Additionally the broker is configured to use basic authentication. The variables
will be:

```sh
$ export AUTH_USERNAME="vault"
$ export AUTH_PASSWORD="broker-secret-password"
```

### Creating a Dev Environment

Note: This section takes a good amount of time (likely, an hour or more) due to downloading a 16gb ISO and VirtualBox taking a while to create an environment.

1. Register for an account at `https://pivotal.io`.
2. Install the `cf` CLI: https://docs.cloudfoundry.org/cf-cli/install-go-cli.html.
3. Ensure you have VirtualBox installed.
4. Ensure you have ~30gb of disk storage available in your local environment for the VirtualBox ISO and the env `cf` will create.
5. Go through installation and configuration of `cf` through the page on the `cf login` step here: https://pivotal.io/platform/pcf-tutorials/getting-started-with-pivotal-cloud-foundry-dev/introduction. Ignore the sample app steps regarding a Spring app.
6. Note that when logging in following the example, the username is literally `user` and the password is literally `pass`, not the username and password you created during registration.
7. Run Vault locally: `$ vault server -dev`. Use the Vault token returned as the `VAULT_TOKEN` value exported above.
8. Make Vault reachable to your VM using ngrok (https://ngrok.com/): `$ ngrok http 8200`. Use the http URL shown as the `VAULT_ADDR` value exported above (example value: "http://4d4053a3.ngrok.io").

### Deploying the Broker

The first step is deploying the broker. This broker can run anywhere including
Cloud Foundry, Kubernetes, Heroku, HashiCorp [Nomad][nomad], or your local
laptop. This example shows running the broker under Cloud Foundry.

First, create a space in which to run the broker:

```shell
$ cf create-space vault-broker
```

Switch over to that space:

```shell
$ cf target -s vault-broker
```

Deploy the vault-broker by cloning this repository:

```shell
$ git clone https://github.com/hashicorp/vault-service-broker
$ cd vault-service-broker
```

And push to Cloud Foundry:

```shell
$ cf push --random-route --no-start
```

- The `--random-route` flag is optional, but it allows us to easily run more
  than one Vault broker if needed, instead of relying on "predictable names".

- The `--no-start` flag is important because we have not supplied the required
  environment variables to our application yet. It will fail to start now.

To configure the broker, provide the following environment variables:

```shell
$ cf set-env vault-broker VAULT_ADDR "$VAULT_ADDR"
$ cf set-env vault-broker VAULT_TOKEN "$VAULT_TOKEN"
$ cf set-env vault-broker SECURITY_USER_NAME "$AUTH_USERNAME"
$ cf set-env vault-broker SECURITY_USER_PASSWORD "$AUTH_PASSWORD"
$ cf restage vault-broker
```

Now that it's configured, start the broker:

```shell
$ cf start vault-broker
```

To verify the HashiCorp Vault broker is running, execute:

```shell
$ cf apps

name           requested state   instances   memory   disk   urls
vault-broker   started           1/1         256M     512M   vault-broker-torpedolike-reexploration.local.pcfdev.io
```

Grab the URL and save it in a variable or copy it to your clipboard - we will need this later.

```shell
export BROKER_URL=$(cf app vault-broker | grep -E -w 'urls:|routes:' | awk '{print $2}')
```

NOTE: Different versions of Cloud Foundry display this information differently. If
the result of the pipeline above is empty, try running `cf app vault-broker` and
look at the output. It is possible that the key has changed again and you'll
need to grep for that instead of `urls:` or `routes:`.

Again, there is no requirement that our broker run under Cloud Foundry - this
could be a URL pointing to any service that hosts the broker, which is just a
Golang HTTP server.

To verify the broker is working as expected, query its catalog:

```shell
$ curl -s "${AUTH_USERNAME}:${AUTH_PASSWORD}@${BROKER_URL}/v2/catalog"
```

The result will be JSON that includes the list of plans for the broker:

```javascript
{
  "services": [
    {
      "id": "0654695e-0760-a1d4-1cad-5dd87b75ed99",
      "name": "hashicorp-vault",
      "description": "HashiCorp Vault Service Broker",
      "bindable": true,
      "tags": [""],
      "plan_updateable": false,
      "plans": [
        {
          "id": "0654695e-0760-a1d4-1cad-5dd87b75ed99.shared",
          "name": "shared",
          "description": "Secure access to Vault's storage and transit backends",
          "free": true
        }
      ]
    }
  ]
}
```

The HashiCorp Vault Service Broker is now running under Cloud Foundry and ready
to receive requests.

### Register the Vault Broker

Before it can bind to services, the broker must be registered with Cloud
Foundry. Remember that there is no requirement that the broker be running under
Cloud Foundry, so we will need to provide the broker registration service the
`BROKER_URL` from above.

To register the broker, an application developer first creates a new space where
they will request the broker access:

```shell
$ cf create-space example
$ cf target -s example
```

Next, register the broker in this space:

```shell
$ cf create-service-broker vault-broker "${AUTH_USERNAME}" "${AUTH_PASSWORD}" "https://${BROKER_URL}" --space-scoped
```

Note: This will allow developers in the current space to access the broker. If
other developers in another space want to access it, they will need to create
their own instance. This is a good starting workflow, but more complex setups
should investigate [allowing access to plans globally or
per-organization][cf-service-acls].

Note: Here we are scoping the broker to a space. In larger installations, it may
be desirable to run a centralized instance of the broker that is accessible to
all spaces in the organization. For more information on this deployment pattern,
please see the notes in the [standard broker](#global-standard-broker) section.

To verify the command worked, query the marketplace. You should see the Vault broker
with a plan of 'default' in addition to any other services you may have access to:

```shell
$ cf marketplace
service           plans             description
hashicorp-vault   shared            HashiCorp Vault Service Broker
# ...
```

### Create a service instance and bind an app

After registering the service in the marketplace, it is now possible to create a
service instance and bind to it.

First, create a service instance using the default plan.  For this example we
will name the service instance 'my-vault':

```shell
$ cf create-service hashicorp-vault shared my-vault
```

With a service instance in place, you are ready to bind an app. Suppose we have
an app called 'my-app'. An example of my-app can be found in the `example` directory
along with instructions on how to deploy it.

```shell
$ cf bind-service my-app my-vault
```

You will need to restage the app to pick up the environment changes.

```shell
$ cf restage my-app
```

When the app starts back up, its `VCAP_SERVICES` environment variable will
contain an entry for the my-vault service.  You can confirm this by checking the
app's environment variables:

```shell
$ cf env my-app
```

The `VCAP_SERVICES` environment variable will have a section similar to
the following:

```json
{
	"hashicorp-vault": [{
		"credentials": {
			"address": "http://ae585ec6.ngrok.io/",
			"auth": {
				"accessor": "kMr3iCSlekSN2d1vpPjbjzUk",
				"token": "s.qgVrPa3eKawwDDkeOSXUaWZq"
			},
			"backends": {
				"generic": [
					"cf/7f1a12a9-4a52-4151-bc96-874380d30182/secret",
					"cf/c4073566-baee-48ae-88e9-7c7c7e0118eb/secret"
				],
				"transit": [
					"cf/7f1a12a9-4a52-4151-bc96-874380d30182/transit",
					"cf/c4073566-baee-48ae-88e9-7c7c7e0118eb/transit"
				]
			},
			"backends_shared": {
				"organization": "cf/8d4b992f-cca3-4876-94e0-e49170eafb67/secret",
				"space": "cf/bdace353-e813-4efb-8122-58b9bd98e3ab/secret"
			}
		},
		"label": "hashicorp-vault",
		"name": "my-vault",
		"plan": "shared",
		"provider": null,
		"syslog_drain_url": null,
		"tags": [],
		"volume_mounts": []
	}]
}
```

The keys of the `credentials` section are as follows:

- `address` - address to the Vault server to make requests against

- `auth.accessor` - token accessor which can be used for logging

- `auth.token` - token to supply with requests to Vault

- `backends.generic` - namespaces in Vault where this token has full CRUD access
  to the static secret storage ("generic") backend, one at the application level
  and the other at the service instance level

- `backends.transit` - namespaces in Vault where this token has full access to
  the transit ("encryption as a service") backend, one at the application level
  and the other at the service instance level

- `backends_shared.organization` - namespace in Vault where this token has
  read-only access to organization-wide data; all instances have read-only
  access to this path, so it can be used to share information across the
  organization.

- `backends_shared.space` - namespace in Vault where this token has read-write
  access to space-wide data; all instances have read-write access to this path,
  so it can be used to share information across the space.

## Internals

### Architecture and Assumptions

To ease in setup and administration, the HashiCorp Vault Service Broker makes a
few assumptions about the Vault setup including:

- The Vault server is already running and is accessible by the broker.

- The Vault server may be used by other applications (it is not exclusively tied
  to Cloud Foundry).

- All instances of an application will _share_ a token. This goes against the
  recommended Vault usage, but this is a limitation of the Cloud Foundry service
  broker model.

- Any Vault operations performed outside of Cloud Foundry will require users to
  rebind their instances.

When a new service instance is provisioned using the broker, it will mount the
following paths:

1. Mount the `generic` backend at `/cf/<organization_id>/secret/`
1. Mount the `generic` backend at `/cf/<space_id>/secret/`
1. Mount the `generic` backend at `/cf/<app_id>/secret/`
1. Mount the `transit` backend at `/cf/<app_id>/transit/`
1. Mount the `generic` backend at `/cf/<instance_id>/secret/`
1. Mount the `transit` backend at `/cf/<instance_id>/transit/`

The mount operation is idempotent, so service instances in the same organization
or space will not re-create the mount. These mount points will be returned to
the application in the secret data, so there is no need to "guess" or
interpolate these strings.

The read-only organization mount allows for sharing secrets organization wide,
and the read-write mount permits sharing secrets between applications in the
same Cloud Foundry Space. The transit backend is mounted to provide
"encryption-as-a-service" on a per-application level.

After mounting is complete, the broker generates a custom policy specific for
this instance which grants the following:

- Read-only access to `"cf/<organization_id>/*"`
- Read-write access to `"cf/<space_id>/*"`
- Full access to `"cf/<instance_id>/*"`

This policy is named `"cf-<instance_id>"` and can be further customized outside
of Cloud Foundry by a Vault administrator.

Next the broker creates a new token role. This role creates a periodic token
with the above policy attached. This **does not** create the token yet, just the
role for generating the token.

When a service instance is bound to an application, the broker performs the
following operations:

- Create a new token against the previous "cf-<instance_id>" role.

- Start a background process to renew this token

- Generate and returning the binding credentials (see above for the schema)

It is important to note that all instances of a Cloud Foundry application
will share the same `vault_token`. This is not the recommended pattern for
using Vault, but it is an existing limitation of the service broker model.

### Unbinding and Deleting

When unbinding from a service or deleting the service broker entirely, the
broker deletes an instance-specific data. For safety, the broker does not delete
an space or organization-specific mounts, even if there are no remaining service
brokers using it.

### Broker Vault Token Permissions

The Cloud Foundry Vault Broker requires a `VAULT_TOKEN` to operate. This token
should have elevated permissions in Vault, as it will be responsible for
generating new mounts and committing data to its internal data structure.

Here is a sample policy to assign to this token:

```hcl
# Manage internal state under "/broker", but since this token is going to
# generate children, it needs full management of the "/cf/*" space
path "/cf/" {
  capabilities = ["list"]
}

path "/cf/*" {
  capabilities = ["create", "read", "update", "delete", "list"]
}

# List all mounts
path "sys/mounts" {
  capabilities = ["read", "list"]
}

# Create mounts under the "/cf/" prefix
path "sys/mounts/cf/*" {
  capabilities = ["create", "update", "delete"]
}

# Create policies with the "cf-*" prefix
path "sys/policies/acl/cf-*" {
  capabilities = ["create", "update", "delete"]
}

# Create token role
path "/auth/token/roles/cf-*" {
  capabilities = ["create", "update", "delete"]
}

# Create tokens from role
path "/auth/token/create/cf-*" {
  capabilities = ["create", "update"]
}

# Revoke tokens by accessor
path "/auth/token/revoke-accessor" {
  capabilities = ["create", "update"]
}
```

Additionally, this token should be a [periodic token][vault-periodic-token]. The
Cloud Foundry Vault Broker will renew this periodic token automatically.

1. Authenticate as a root token or user with sudo privilege in Vault (this is
  required to create a periodic token):

  ```shell
  $ vault auth <token>
  ```

1. Create the policy specific for the broker:

  ```shell
  $ vault write cf-broker cf-broker.hcl
  ```

1. Create a periodic token

  ```shell
  $ vault token-create -period="30m" -orphan -policy=cf-broker
  ```

  Grab the value for "token" and store it somewhere safe for now - you will need
  this when configuring the HashiCorp Vault Service Broker.

### Service Broker Configuration

The service broker is designed to be configured using environment variables. It
currently recognizes the following.

- `SERVICE_DESCRIPTION` (default: "HashiCorp Vault Service Broker") -
  description of the service to show in the marketplace

- `SERVICE_ID` (default: "0654695e-0760-a1d4-1cad-5dd87b75ed99") - UUID of the
  broker

- `SERVICE_NAME` (default: "hashicorp-vault") - name of the service to show in
  the marketplace

- `SERVICE_TAGS` (default: none) - comma-separated list of tags for the service

- `PLAN_NAME` (default: "shared") - the name of the plan in the marketplace

- `PLAN_DESCRIPTION` (default: "Secure access to Vault's storage and transit backends") - description of the plan in the marketplace

- `PORT` (default: "8000") - port to bind and listen on as the server (broker)

- `VAULT_ADDR` (default: "https://127.0.0.1:8200") - address to the Vault server

- `VAULT_ADVERTISE_ADDR` (default: "$VAULT_ADDR") - address to advertise to
  clients as Vault's address. This defaults to the value supplied for
  `VAULT_ADDR`, but can be overridden. This is most useful when the broker
  communicates to Vault on a local subnet, but clients communicate through a
  public subnet.

- `VAULT_RENEW` (default: true) - enable renewal of the token provided to Vault.
  The token given to Vault is assumed to be a periodic token, and the broker
  will automatically renew it to prevent it from expiring. If an out-of-band
  process is managing the renewal, disable this by setting it to "false".

- `VAULT_TOKEN` (default: none) - token to authenticate the broker to Vault.
  This token should have permission to mount and unmount backends, read, list,
  and delete paths, and create tokens with role permissions. Please see the
  [Vault Token Permissions](#vault-token-permissions) section for more
  information on the requirements for this token.
  
- `VAULT_NAMESPACE` - (default: none) - namespace to use for all calls within Vault

- `SECURITY_USER_NAME` - (default: none) - username for basic auth

- `SECURITY_USER_PASSWORD` - (default: none) - password for basic auth

### Providing Configuration Through CredHub

Any of the Vault Service Broker's environment variables can be set through CredHub. 
Authenticating to CredHub is typically done by using a UAA server. To configure this,
the following environment variables must be set:

- `CREDHUB_URL` (default: none) - CredHub's base URL (ex. "https://example.com")
- `UAA_ENDPOINT` (default: none) - UAA's base URL (ex. "http://localhost:8080/uaa")
- `UAA_CLIENT_NAME` (default: none) - Client name to use when gaining CredHub auth token through UAA
- `UAA_CLIENT_SECRET` (default: none) - Client secret to use when gaining CredHub auth token through UAA

The following optional parameters are also available:

- `UAA_CA_CERTS` (default: none) - CA certs to use when authenticating to UAA
- `UAA_SKIP_VERIFICATION` (default: false) - Skip verifying certificates when calling UAA
- `UAA_INSECURE_ALLOW_ANY_SIGNING_METHOD` (default: false) - Allow any signing method when verifying UAA certs

Once this is set, Vault will check CredHub for all environment variables listed above.
All variables must be prefixed with "VAULT_SERVICE_BROKER_". For example:

- `VAULT_SERVICE_BROKER_SECURITY_USER_PASSWORD`: "$ecure_pa$$w0rd"
- `VAULT_SERVICE_BROKER_VAULT_RENEW`: "false"
- `VAULT_SERVICE_BROKER_SERVICE_TAGS`: "production,security"

Please keep the following in mind:
- You may set configuration through both CredHub and environment variables
- CredHub is preferred, so if a variable exists in both places, the CredHub value will prevail
- The values for the CredHub variables must be given as strings in the same format as you would an environment variable

### Granting Access to Other Paths

The service broker has an opinionated setup of policies and mounts to provide a
simplified user experience for getting started which matches the organizational
model of Cloud Foundry. However an application may require access to existing
data or backends.

Once the broker creates the policy for a service id `cf-<instance_id>` that
policy may be modified by a user with permissions in Vault to add additional
capabilities. The default policy can be discovered by reading it:

```sh
$ vault policy read cf-<instance_id>
# ...
```

Append any additional rules to the end.

### Global Standard Broker

The default configuration and examples above use a "space scoped" broker. For
larger installations, it may be desirable to run a centralized instance of the
broker that is accessible to all spaces and applications in the system. This is
generally called a "standard broker" in Cloud Foundry terminology. To deploy a
global broker, an admin must add, publish, and permit access as follows.

First, create the broker. Note the lack of `--space-scoped` as compared to the
previous commands.

```shell
$ cf create-service-broker vault-broker "${AUTH_USERNAME}" "${AUTH_PASSWORD}" "https://${BROKER_URL}"
```

To verify the broker was created:

```shell
$ cf service-access
# ...

broker: vault-broker
   service           plan     access   orgs
   hashicorp-vault   shared   none
```

Notice the access is listed as "none". This broker will not appear in the
marketplace until activated by an admin. To activate, run:

```shell
$ cf enable-service-access hashicorp-vault
```

It is possible to restrict the access to particular organizations or plans.

```shell
$ cf enable-service-access hashicorp-vault -o my-org
```

Verify the plan is enabled:

```shell
$ cf service-access
# ...

broker: vault-broker
   service           plan     access   orgs
   hashicorp-vault   shared   all
```

And check in the marketplace

```shell
$ cf marketplace
service           plans             description
hashicorp-vault   shared            HashiCorp Vault Service Broker
# ...
```

Now all spaces have access to this centralized broker!

## Contributing

1. Clone the repo
1. Make changes on a branch
1. Test changes
1. Submit a Pull Request to GitHub

[cf-service-acls]: https://docs.cloudfoundry.org/services/access-control.html "Cloud Foundry Service ACLs"
[nomad]: https://www.nomadproject.io/ "Nomad by HashiCorp"
[vault]: https://www.vaultproject.io/ "Vault by HashiCorp"
[vault-periodic-token]: https://www.vaultproject.io/docs/concepts/tokens.html#token-time-to-live-periodic-tokens-and-explicit-max-ttls "Vault Periodic Tokens"


CHANGELOG.md
# Cloud Foundry HashiCorp Vault Broker Changelog

## v0.6.0 (Aug 11, 2022)
CHANGES:
- [#65](https://github.com/hashicorp/vault-service-broker/pull/65) migrated to go modules, use go 1.18 in tests, and updated to vault/api@v1.7.2

IMPROVEMENTS:
- [#66](https://github.com/hashicorp/vault-service-broker/pull/66) allow environment variables like `VAULT_SKIP_VERIFY` to be picked up by the Vault client

## v0.5.4 (Dec 2, 2020)
BUG FIXES:
- [#57](https://github.com/hashicorp/vault-service-broker/pull/57) fixes an issue where the broker sends too many requests to Vault during startup

## v0.5.3 (May 13, 2019)
BUG FIXES:
- [#49](https://github.com/hashicorp/vault-service-broker/pull/49) fixes an issue during upgrades when broker version <= 0.2.0
- [#48](https://github.com/hashicorp/vault-service-broker/pull/48) continue the unbind when the token has already expired or revoked out of band

## v0.5.2 (April 23, 2019)
BUG FIXES:
- [#47](https://github.com/hashicorp/vault-service-broker/pull/47) prevent errors when PCF does not provide an application id

## v0.5.1 (February 15, 2019)
BUG FIXES:
- [#43](https://github.com/hashicorp/vault-service-broker/pull/43) resolves a bug where it was impossible to unbind if a token had been deleted out-of-band

## v0.5.0 (January 30, 2019)
FEATURES:
- [#42](https://github.com/hashicorp/vault-service-broker/pull/42) adds support for a more polished-looking Marketplace UI

BUG FIXES:
- [#41](https://github.com/hashicorp/vault-service-broker/pull/41) solves multiple minor issues that caused black box testing to fail using `$ cf dev` locally - namely, it updates the go buildpack, the version of go used by cf for building, and deletes two unnecessary files that were symlinked outside the repo

## v0.4.0 (January 29, 2019)
FEATURES:
- [#33](https://github.com/hashicorp/vault-service-broker/pull/33) adds support for a broker-level namespace setting
- [#39](https://github.com/hashicorp/vault-service-broker/pull/39) adds support for application-level secrets engines but also contains a **breaking change** in the format of the JSON returned by the `Bind` call so should be adopted with caution

## v0.3.0 (January 23, 2019)

FEATURES:
- [#37](https://github.com/hashicorp/vault-service-broker/pull/37) adds support for all configuration variables to be provided through CredHub with UAA for CredHub authentication

IMPROVEMENTS:
- Adds functional tests documenting current behavior of the Vault Service Broker
- Adds Travis CI builds

BUG FIXES:
- Resolves [#25](https://github.com/hashicorp/vault-service-broker/issues/25), an extraneous error renewing the Vault token on startup


## v0.2.0 (July 24, 2017)

- Initial Release
- Implements the Bind, Unbind, Provision, and Deprovision calls for the Open Service Broker API


main.go
package main

import (
	"errors"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"reflect"
	"strconv"
	"strings"
	"syscall"

	"code.cloudfoundry.org/clock"
	"code.cloudfoundry.org/lager"
	uaa "code.cloudfoundry.org/uaa-go-client"
	uaaconf "code.cloudfoundry.org/uaa-go-client/config"
	credhub "github.com/cloudfoundry-community/go-credhub"
	"github.com/hashicorp/go-cleanhttp"
	"github.com/hashicorp/vault/api"
	"github.com/kelseyhightower/envconfig"
	"github.com/pivotal-cf/brokerapi"
)

func main() {
	// Setup the logger - intentionally do not log date or time because it will
	// be prefixed in the log output by CF.
	logger := log.New(os.Stdout, "", 0)
	cfLogger := lager.NewLogger("vault-broker")

	config, err := parseConfig(cfLogger)
	if err != nil {
		logger.Fatal("[ERR] failed to read configuration", err)
	}

	// Setup the vault client
	vaultClientConfig := api.DefaultConfig()

	vaultClient, err := api.NewClient(vaultClientConfig)
	if err != nil {
		logger.Fatal("[ERR] failed to create vault api client", err)
	}

	vaultClient.SetAddress(config.VaultAddr)
	vaultClient.SetToken(config.VaultToken)
	if config.VaultNamespace != "" {
		vaultClient.SetNamespace(config.VaultNamespace)
	}

	// Setup the broker
	broker := &Broker{
		log:         logger,
		vaultClient: vaultClient,

		serviceID:          config.ServiceID,
		serviceName:        config.ServiceName,
		serviceDescription: config.ServiceDescription,
		serviceTags:        config.ServiceTags,

		planName:         config.PlanName,
		planDescription:  config.PlanDescription,
		planMetadataName: config.PlanMetadataName,
		planBullets:      config.PlanBullets,

		displayName:         config.DisplayName,
		imageUrl:            config.ImageUrl.String(),
		longDescription:     config.LongDescription,
		providerDisplayName: config.ProviderDisplayName,
		documentationUrl:    config.DocumentationUrl,
		supportUrl:          config.SupportUrl,

		vaultAdvertiseAddr: config.VaultAdvertiseAddr,
		vaultRenewToken:    config.VaultRenew,
	}
	if err := broker.Start(); err != nil {
		logger.Fatalf("[ERR] failed to start broker: %s", err)
	}

	// Parse the broker credentials
	creds := brokerapi.BrokerCredentials{
		Username: config.SecurityUserName,
		Password: config.SecurityUserPassword,
	}

	// Setup the HTTP handler
	handler := brokerapi.New(broker, cfLogger, creds)

	// Listen to incoming connection
	serverCh := make(chan struct{}, 1)
	go func() {
		logger.Printf("[INFO] starting server on %s", config.Port)
		if err := http.ListenAndServe(config.Port, handler); err != nil {
			logger.Fatalf("[ERR] server exited with: %s", err)
		}
		close(serverCh)
	}()

	signalCh := make(chan os.Signal, 1)
	signal.Notify(signalCh, syscall.SIGTERM, syscall.SIGINT)

	select {
	case <-serverCh:
	case s := <-signalCh:
		logger.Printf("[INFO] received signal %s", s)
	}

	if err := broker.Stop(); err != nil {
		logger.Fatalf("[ERR] faild to stop broker: %s", err)
	}

	os.Exit(0)
}

// normalizeAddr takes a string that represents a URL and ensures it has a
// scheme (defaulting to https), and ensures the path ends in a trailing slash.
func normalizeAddr(s string) string {
	if s == "" {
		return s
	}

	u, err := url.Parse(s)
	if err != nil {
		return s
	}

	if u.Scheme == "" {
		u.Scheme = "https"
	}

	if strings.Contains(u.Scheme, ".") {
		u.Host = u.Scheme
		if u.Opaque != "" {
			u.Host = u.Host + ":" + u.Opaque
			u.Opaque = ""
		}
		u.Scheme = "https"
	}

	if u.Host == "" {
		split := strings.SplitN(u.Path, "/", 2)
		switch len(split) {
		case 0:
		case 1:
			u.Host = split[0]
			u.Path = "/"
		case 2:
			u.Host = split[0]
			u.Path = split[1]
		}
	}

	u.Path = strings.TrimRight(u.Path, "/") + "/"

	return u.String()
}

func parseConfig(cfLogger lager.Logger) (*Configuration, error) {
	config := &Configuration{}
	if err := envconfig.Process("", config); err != nil {
		return nil, err
	}
	if config.CredhubURL != "" {
		if err := credhubProcess(cfLogger, "VAULT_SERVICE_BROKER_", config); err != nil {
			return nil, err
		}
	}
	if err := config.Validate(); err != nil {
		return nil, err
	}
	return config, nil
}

type Configuration struct {
	// Required
	SecurityUserName     string `envconfig:"security_user_name"`
	SecurityUserPassword string `envconfig:"security_user_password"`
	VaultToken           string `envconfig:"vault_token"`

	// Optional, for using CredHub
	CredhubURL                       string `envconfig:"credhub_url"`
	UAAEndpoint                      string `envconfig:"uaa_endpoint"`
	UAAClientName                    string `envconfig:"uaa_client_name"`
	UAAClientSecret                  string `envconfig:"uaa_client_secret"`
	UAACACerts                       string `envconfig:"uaa_ca_certs"`
	UAASkipVerification              bool   `envconfig:"uaa_skip_verification"`
	UAAInsecureAllowAnySigningMethod bool   `envconfig:"uaa_insecure_allow_any_signing_method"`

	// Also optional
	Port               string `envconfig:"port" default:":8000"`
	ServiceID          string `envconfig:"service_id" default:"0654695e-0760-a1d4-1cad-5dd87b75ed99"`
	VaultAddr          string `envconfig:"vault_addr" default:"https://127.0.0.1:8200"`
	VaultAdvertiseAddr string `envconfig:"vault_advertise_addr"`
	VaultNamespace     string `envconfig:"vault_namespace"`
	ServiceName        string `envconfig:"service_name" default:"hashicorp-vault"`
	ServiceDescription string `envconfig:"service_description" default:"HashiCorp Vault Service Broker"`
	PlanName           string `envconfig:"plan_name" default:"shared"`
	PlanDescription    string `envconfig:"plan_description" default:"Secure access to Vault's storage and transit backends"`

	PlanMetadataName    string          `envconfig:"plan_metadata_name" default:"Architecture and Assumptions"`
	PlanBullets         []string        `envconfig:"plan_bullets" default:"The Vault server is already running and is accessible by the broker.,The Vault server may be used by other applications (it is not exclusively tied to Cloud Foundry).,All instances of an application will share a token. This goes against the recommended Vault usage. This is a limitation of the Cloud Foundry service broker model.,Any Vault operations performed outside of Cloud Foundry will require users to rebind their instances."`
	DisplayName         string          `envconfig:"display_name" default:"Vault for PCF"`
	ImageUrl            *ImageDefaulter `envconfig:"image_url" default:"unset"`
	LongDescription     string          `envconfig:"long_description" default:"The official HashiCorp Vault broker integration to the Open Service Broker API. This service broker provides support for secure secret storage and encryption-as-a-service to HashiCorp Vault."`
	ProviderDisplayName string          `envconfig:"provider_display_name" default:"HashiCorp"`
	DocumentationUrl    string          `envconfig:"documentation_url" default:"https://www.vaultproject.io/"`
	SupportUrl          string          `envconfig:"support_url" default:"https://support.hashicorp.com/"`

	ServiceTags []string `envconfig:"service_tags"`
	VaultRenew  bool     `envconfig:"vault_renew" default:"true"`
}

func (c *Configuration) Validate() error {
	// Ensure required parameters were provided
	if c.SecurityUserName == "" {
		return errors.New("missing SECURITY_USER_NAME")
	}
	if c.SecurityUserPassword == "" {
		return errors.New("missing SECURITY_USER_PASSWORD")
	}
	if c.VaultToken == "" {
		return errors.New("missing VAULT_TOKEN")
	}

	// If these values aren't perfect, we can fix them
	if !strings.HasPrefix(c.Port, ":") {
		c.Port = ":" + c.Port
	}
	if c.VaultAdvertiseAddr == "" {
		c.VaultAdvertiseAddr = c.VaultAddr
	}
	c.VaultAddr = normalizeAddr(c.VaultAddr)
	c.VaultAdvertiseAddr = normalizeAddr(c.VaultAdvertiseAddr)
	return nil
}

// credhubProcess iterates over the names of variables as set in the `envconfig` tag
// on the Configuration. It prepends them with "prefix" and then looks
// in Credhub to see if they exist. If they do and they have a value, the Configuration
// is updated with that value for that field.
func credhubProcess(cfLogger lager.Logger, prefix string, config *Configuration) error {
	uaaConf := &uaaconf.Config{
		ClientName:                    config.UAAClientName,
		ClientSecret:                  config.UAAClientSecret,
		UaaEndpoint:                   config.UAAEndpoint,
		SkipVerification:              config.UAASkipVerification,
		CACerts:                       config.UAACACerts,
		InsecureAllowAnySigningMethod: config.UAAInsecureAllowAnySigningMethod,
	}
	uaaClient, err := uaa.NewClient(cfLogger, uaaConf, clock.NewClock())
	if err != nil {
		return err
	}
	credhubClient := credhub.New(config.CredhubURL, credhub.NewUAAAuthClient(cleanhttp.DefaultClient(), uaaClient))

	// Pull the "envconfig" field name from each field and look for it in Credhub
	configTypeInfo := reflect.TypeOf(*config)
	settableConfig := reflect.ValueOf(config).Elem()

	for i := 0; i < configTypeInfo.NumField(); i++ {
		fieldTypeInfo := configTypeInfo.Field(i)
		credhubName := prefix + strings.ToUpper(fieldTypeInfo.Tag.Get("envconfig"))

		latest, err := credhubClient.GetLatestByName(credhubName)
		if err != nil && !strings.Contains(strings.ToLower(err.Error()), "not found") {
			return err
		}
		if latest == nil {
			// This key doesn't exist in Credhub
			continue
		}
		settingValue, ok := latest.Value.(string)
		if !ok {
			return fmt.Errorf("we only support credhub values as bash-like string values, but received %s as a %s", credhubName, reflect.TypeOf(latest.Value))
		}
		if settingValue == "" {
			// The value for this key isn't set in Credhub
			continue
		}

		// Update the value for this field with Credhub's value
		settableField := settableConfig.Field(i)
		switch fieldTypeInfo.Type.Kind() {
		case reflect.Bool:
			asBool, err := strconv.ParseBool(settingValue)
			if err != nil {
				return fmt.Errorf("error parsing bool %s: %s", credhubName, err)
			}
			settableField.SetBool(asBool)
		case reflect.String:
			settableField.SetString(settingValue)
		case reflect.Slice:
			settableField.Set(reflect.ValueOf(strings.Split(settingValue, ",")))
		default:
			return fmt.Errorf("unsupported type of %s for %s", fieldTypeInfo.Type.Kind(), credhubName)
		}
	}
	return nil
}

type ImageDefaulter struct {
	Image string
}

func (d *ImageDefaulter) Decode(image string) error {
	d.Image = image
	if d.Image == "unset" {
		def, err := ioutil.ReadFile("dflt_img_url.txt")
		if err != nil {
			return err
		}
		d.Image = fmt.Sprintf("%s", def)
	}
	return nil
}

func (d *ImageDefaulter) String() string {
	return d.Image
}
